time/ptp.md

```diff
@@ -0,0 +1,218 @@
+
+#### PTP4L configuration file
+
+Default ptp4l configuration file (ptp4l.conf) will be used when starting ptp4l
+service, user doesn't need to specify ptp4l.conf in `ptp4lOpts`. The content
+of ptp4l.conf is using default file delivered in linuxptp package. This
+configuration file cannot be changed via CRDs exposed in this proposal. However,
+some of ptp4l configuration options in ptp4l.conf can be overwrited by
+specifying `ptp4lOpts` in `NodePTPCfg` custom resource.
+
+#### Redundancy
+
+##### PTP port redundancy
+
+Ptp4l supports using bonding interface in active-backup mode. It uses the
+active interface from bond as PTP clock and can switch to another active
+interface in case of a failover. Creation of bonding interface(e.g. bond0)
+is not supported in this proposal, to use bond interface as PTP device,
+user needs to create it manually(for example, on RHEL nodes).
+
+##### PTP and NTP redundancy
+
+Chrony is deployed and enabled by default during OpenShift installation process.
+This proposal doesn't support timemaster service configuration. Use of PTP will
+require disabling NTP(Chrony) first. No redundancy is provided between NTP and
+PTP in this proposal.
+
+### Risks and Mitigations
+
+In case of failure in linuxptp configuration, OpenShift nodes will be left as
+no time source to sync, resulting in potential time disorder among nodes.
+This can be resolved by providing PTP and NTP redundancy via timemaster service
+which automatically rolls back to use default NTP time source if configured.
+
+PTP device working in UDP transport mode requires IP configuration on network
+interface, if the network interface is also used by default openshift-sdn
+plugin, it may destroy network connection established by openshift-sdn. This
+should be documented as a risk.
+
+## Design Details
+
+### Test Plan
+
+- Tests will not be conducted against real PTP grandmaster clock
+- Functional tests will be implemented
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Linuxptp can be installed via container image
+- Linuxptp services can be configured via CRDs
+- End user documentation on how to interact with PTP DaemonSet
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+#### Tech Preview -> GA
+
+- Have an opeartor to manage PTP CRDs
+- Support advanced configuration of linuxptp such as timemaster
+- Support all PTP network transport modes
+- Provide redundancy support for PTP, both port and time source redundancy
+- Support configuration of all PTP Clock types
+- Measure PTP time accuracy in real environment
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+PTP runs as a separate DaemonSet.
+
+## Implementation History
+
+Version 4.3
+
+Tech Preview
+
+## Infrastructure Needed [optional]
+
+This requires a github repo be created under openshift org to hold PTP
+DaemonSet code. The name of this repo is `ptp-daemon`.
```

Comment:
As mentioned above, there will be the need of an operator managing the PTP daemonset. Then, there will be two different repos? one with the ptp-daemon and one with the ptp-operator?

Reply:
Yes, the idea is to first implement a PTP daemon in 4.3 as Tech Preview, and have an operator to manage this daemon and exposes all CRDs mentioned in proposal. Operator will need a separate repo under openshift github org (this is not strictly required, as you can see, SR-IOV Operator uses one repo to hold code for both daemon and operator).

Reply:
Yeah, that's why I mentioned it. I think the trend in OpenShift is to name the repos according to the operator, even if they contains the code of the actual application that is being "operated".

Reply:
After looking at this some more it seems like there are more examples of separate repos for operator and operands, ie: image-registry, kubernetes-apiserver. So it seems there's no norm established here so lets just go with separate repos as initially proposed.